### ISSUE TITLE
examples needed

### ISSUE THREAD

**Issue by maceip** [microsoftgraph/msgraph-sdk-go-core#19]
i'm new to go but after playing with this sdk for a bit, getting basic things to work took a long time and I gave up.

I gave up on azidentity.NewClientCertificateCredential (does it need a fs path to a cert or a []*x509.Certificate?)
I couldnt figure out how to set the right options for .IdentityProvidersRequestBuilderGetOption
I couldn't figure out how to parse responses:
```
result: &{{map[error:map[code:0xc000099c40 innerError:0xc0001bcf20 message:0xc000099d50]] <nil>} <nil>}
result: &{map[error:map[code:0xc000012e60 innerError:0xc0000130e0 message:0xc000012f00]] <nil> []}
cannot range over result (type *graph.IdentityProviderBase)
cannot range over result2 (type *"github.com/microsoftgraph/msgraph-beta-sdk-go/identity/identityproviders".IdentityProvidersResponse)
```

the only example I could find used result.GetId()) when calling Drive()

Can we get an example of how to authenticate with a certificate, set options correctly, and work with responses?

Your docs currently show this for most methods:
```
//THE GO SDK IS IN PREVIEW. NON-PRODUCTION USE ONLY
graphClient := msgraphsdk.NewGraphServiceClient(requestAdapter)
result, err := graphClient.Identity().IdentityProviders().Get(options)
```

what are the options? how do I parse the results?

**Comment by baywet** [comment]
Hi @maceip 
Thanks for reaching out and for the very valuable feedback.

## Options being referred to in the snippets when undefined

The fact that options are present in snippets (variable is referred to but not present), is actually a bug of the snippet generation code [I have fixed](https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/788) but the documentation hasn't been refreshed since. (Thanksgiving in the US) I expect this change will roll out in the public documentation some time next week, maybe the following one.

## What should I provide for options?

nil if you don't have any request body, header, query parameter, request option to set. An option value if you want to set any of those things. Your IDE should allow you to navigate to the Get method definition (usually F12) and from there navigate to the struct definition so you can explore the type.
Please let us know if you think we could improve the discovery experience.
Alternatively, you can check [the online go docs](https://pkg.go.dev/github.com/microsoftgraph/msgraph-sdk-go@v0.3.0/identity/identityproviders#IdentityProvidersRequestBuilderGetOptions)

## How can I setup my authentication?

We do have extensive documentation [on this page](https://docs.microsoft.com/en-us/graph/sdks/choose-authentication-providers?tabs=Go), let us know if we're missing anything/if we could improve things. (besides the paragraph below)

## Client certificate option missing from the docs

That's very good feedback, I think we're missing that option across languages for [this page](https://docs.microsoft.com/en-us/graph/sdks/choose-authentication-providers?tabs=Go#client-credentials-provider) CC @jasonjoh to schedule an update of the page.

## Client certificate does not have documentation in the golang docs

Additionally, [this page](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#NewClientCertificateCredential) is lacking an example. Even the [azure sdk repo](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/azidentity) doesn't seem to provide more examples. Could you [create a new issue](https://github.com/Azure/azure-sdk-for-go/issues) in their repo to ask for an example for that credential please?

I believe the assumption is that azure SDK users will leverage IO methods and [x509.ParseCertificates](https://pkg.go.dev/crypto/x509#ParseCertificates) to get the value for that argument.

Hopefully this helps. Let us know if you have further questions/comments.

**Comment by maceip** [comment]
Thanks for your responses.

Somehow on the go pkg doc I saw:

cred, err := azidentity.NewClientCertificateCredential("<tenant ID>",
"<client ID>", "<certificate path>",
&ClientCertificateCredentialOptions{Password: "<certificate
password>"})
if err != nil {
	// process error
}

Which didn't work. The one you linked shows the correct call arguments.

I also wanted to stress the importance of giving simple examples of parsing
responses. I couldn't find anything aside from the blog post that uses
drive getid()

E.g., what's the  preferred way to iterate over all the responses?


Also what auth scopes should be used in a real-world example beyond
user.read:

https://github.com/microsoft/kiota/tree/main/authentication/go/azure

**Comment by baywet** [comment]
If you could provide the link to the pkg doc you found that example, it'd be helpful for us to go fix it.

Are you referring to [this blog post](https://devblogs.microsoft.com/microsoft365dev/go-developers-can-now-access-microsoft-graph-with-streamlined-authentication-and-more/) ?

For permissions, the public documentation has a permissions table for every endpoint that's available on the API. See [this example](https://docs.microsoft.com/en-us/graph/api/user-list?view=graph-rest-1.0&tabs=go#permissions) That's something we document for the whole API, and don't duplicate on a per language basis because the API is too large to maintain that information in sync across a dozen languages. We have had design sessions to see whether we could include a link to the API documentation in the doc comment or include a list of the required scopes in the documentation comments, but this is blocked by upstream technical requirements at the moment.

For the iteration question, are you referring to endpoints that return a collection of items? I just answered a [similar question](https://github.com/microsoftgraph/msgraph-sdk-go/issues/30) today.

**Comment by maceip** [comment]
Here is the example I used to try to get cert auth to work:
https://github.com/Azure/azure-sdk-for-go/wiki/Azure-Identity-Examples#authenticating-a-service-principal-with-a-client-certificate

And above I didn't mean examples of how to iterate/paginate (but that would
be nice) but in general how to handle responses from the client across
different endpoints. e.g,
the type returned from
graphClient.Directory().FederationConfigurationsById(identityProviderBaseId).Get(options)
is (type *graph.IdentityProviderBase) while a call to
graphClient.Identity().IdentityProviders().Get(ioptions)
is (type *"
github.com/microsoftgraph/msgraph-beta-sdk-go/identity/identityproviders
".IdentityProvidersResponse)

How can I simply just print out the json/text for those?

Also I would suggest you add a go-specific auth scopes example (since kiota
is sprawling) showing how most go authors will likely call it (i'm guessing
they will almost always use a client credentials provider, and this page is
overkill for most golang usecases
<https://docs.microsoft.com/en-us/graph/sdks/choose-authentication-providers?tabs=Go>
):

NewAzureIdentityAuthenticationProviderWithScopes(cred,[]string{"
https://graph.microsoft.com/.default"})


Which I had to google around to figure out.

**Comment by baywet** [comment]
Thanks for providing the URL, I did create an issue for the azure team to update the wiki https://github.com/Azure/azure-sdk-for-go/issues/16296

Responses are automatically deserialized to the provided models by the SDK. This is part of the value proposition of the SDK. It is technically possible to accept the native response object directly if you'd like to, look at the response handler on the options (the one we were talking about at the beginning of this thread).
Can you expand on why you'd like to access the JSON response directly instead of a deserialized object?

The `https://graph.microsoft.com/.default` value is the default value that is passed if you call `NewAzureIdentityAuthenticationProvider` instead of `NewAzureIdentityAuthenticationProviderWithScopes`. But I can see how this isn't pointed out in our documentation currently. @jasonjoh to add a comment to suggest the alternative in the samples.

Keep the feedback coming, this is helpful, especially at such an early stage of the SDK's lifecycle.

**Comment by maceip** [comment]
When I call it without scopes it gives me an error so I have to call it
with scopes using that default string and sends my service principle
already has a roll attached. It seems to work?

I'm very new to go and I don't actually want to print out the return values
from the client. I just am looking for examples on how to use it. I have no
idea how to access the fields. Maybe I'm just missing something or not
communicating effectively here.

**Comment by baywet** [comment]
I've just realized there's a bug in the logic for the default initialization of the scopes, the scopes for the token options will never be initialized... https://github.com/microsoft/kiota/blob/9f2719104e807aae74a752c01e9e0c4f7a410794/authentication/go/azure/azure_identity_authentication_provider.go#L35

Let me put a pull request together to address this.

As per your question about accessing the data, if we take your initial example

```Golang
graphClient := msgraphsdk.NewGraphServiceClient(requestAdapter)
result, err := graphClient.Identity().IdentityProviders().Get(nil)

// check err ...

firstValue := result.GetValue()[0]
// give you the first value that was returned, you can also loop on that
nextLink := result.GetNextLink()
// give you a next link if there's a second page to query (see the other issue I linked earlier)
```

Hopefully that clarifies some of the usage patterns?
We also have a pull request coming for the documentation that also explains some of the patterns https://github.com/microsoftgraph/microsoft-graph-docs/pull/14998, you might want to check it out.

**Comment by baywet** [comment]
Added a pull request to address the default initialization issue https://github.com/microsoft/kiota/pull/892

**Comment by maceip** [comment]
Thanks for this, super helpful. I see Jason's branch of go snippets but
none of them process the result. I'd suggest printing the returned fields
to stdout.

Doing result.GetValue()[0] seems about as janky as dumping the json.

**Comment by baywet** [comment]
The result is a collection of items, by conventions in OData collections are wrapped in a object with a next link property and a value property, the Get part is convention for getters in Go, the [0] is indexing into the collection, then you have access to the full identity provider object. I'm not sure what's janky about that?

Unless you meant to get a single object by its Id?

**Comment by maceip** [comment]
Things like this:


result.GetValue undefined (type *graph.IdentityProviderBase has no field or
method GetValue)

**Comment by maceip** [comment]
Giving up again for the night

Trying to get :

https://github.com/microsoftgraph/msgraph-sdk-go/issues/30

To work:

cannot use result.GetValue() (type []graph.User) as type graph.User in
append
./aad.go:44:28: users.NewUsersRequestBuilder undefined (type []graph.User
has no field or method NewUsersRequestBuilder)

**Comment by baywet** [comment]
ok, let's backup for a second. I do think there's a communication issue getting in the way here.
Can you describe at a high level what you are trying to achieve? e.g. "I'm trying to list out all the identity providers"  or "I'm trying to get a single identity provider via its id"

**Comment by maceip** [comment]
At this point I'm just I'm trying to run your examples and jlprint out
values as a starting point.

For List() endpoints, can you have your examples print out the values ?

**Comment by baywet** [comment]
```Golang
graphClient := msgraphsdk.NewGraphServiceClient(requestAdapter)
result, err := graphClient.Identity().IdentityProviders().Get(nil)

// check err ...

items := result.GetValue()

// prints all the client ids for the first page of identity providers
for _, item := range items {
    fmt.Println("Client id is:", *(item.GetClientId()))
}
```

Also, the pull request fix the default value for the `NewAzureIdentityAuthenticationProvider` has been merged. You should be able to use it now after running `go get -u projectname`.

**Comment by maceip** [comment]
🙏
